### PR TITLE
refactor(rest): improve inspect api

### PR DESCRIPTION
**Pull request by BlackHole1**
Add the correct cpus, memory, devices

**Test Code:**
```golang
package main

import (
	"os"
	"sync"

	"github.com/crc-org/vfkit/pkg/config"
	"github.com/crc-org/vfkit/pkg/rest/define"
	"github.com/gin-gonic/gin"
)

var (
	once            sync.Once
	devicesResponse define.DevicesResponse
)

func devicesToResp(devices []config.VirtioDevice) define.DevicesResponse {
	once.Do(func() {
		for _, dev := range devices {
			switch d := dev.(type) {
			case *config.USBMassStorage:
				devicesResponse.USBMassStorage = append(devicesResponse.USBMassStorage, *d)
			case *config.VirtioBlk:
				devicesResponse.Blk = append(devicesResponse.Blk, *d)
			case *config.RosettaShare:
				devicesResponse.Rosetta = *d
			case *config.NVMExpressController:
				devicesResponse.NVMe = append(devicesResponse.NVMe, *d)
			case *config.VirtioFs:
				devicesResponse.FS = append(devicesResponse.FS, *d)
			case *config.VirtioNet:
				n := define.VirtioNetResponse{
					Nat:            d.Nat,
					MacAddress:     d.MacAddress.String(),
					UnixSocketPath: d.UnixSocketPath,
				}

				if d.Socket != nil {
					n.Fd = int(d.Socket.Fd())
				}

				devicesResponse.Net = append(devicesResponse.Net, n)
			case *config.VirtioRng:
				devicesResponse.Rng = true
			case *config.VirtioSerial:
				devicesResponse.Serial = *d
			case *config.VirtioVsock:
				devicesResponse.Vsock = append(devicesResponse.Vsock, *d)
			case *config.VirtioInput:
				devicesResponse.Input = append(devicesResponse.Input, *d)
			case *config.VirtioGPU:
				devicesResponse.GPU = append(devicesResponse.GPU, *d)
			}
		}
	})

	return devicesResponse
}

func Inspect(devices []config.VirtioDevice) define.InspectResponse {
	ii := define.InspectResponse{
		Devices: devicesToResp(devices),
	}
	return ii
}

func main() {
	eng := gin.Default()
	eng.GET("/inspect", func(c *gin.Context) {
		var devs []config.VirtioDevice

		usb, _ := func() (config.VirtioDevice, error) { return config.USBMassStorageNew("/usb1") }()
		usb2, _ := func() (config.VirtioDevice, error) { return config.USBMassStorageNew("/usb2") }()
		blk, _ := func() (config.VirtioDevice, error) { return config.VirtioBlkNew("/a") }()
		blk2 := &config.VirtioBlk{
			StorageConfig: config.StorageConfig{
				ImagePath: "/b",
				DevName:   "virtio-blk",
			},
			DeviceIdentifier: "/vdd",
		}
		rosetta, _ := func() (config.VirtioDevice, error) { return config.RosettaShareNew("/rosetta") }()
		nvme, _ := func() (config.VirtioDevice, error) { return config.NVMExpressControllerNew("/nvme1") }()
		nvme2, _ := func() (config.VirtioDevice, error) { return config.NVMExpressControllerNew("/nvme2") }()
		fs, _ := func() (config.VirtioDevice, error) { return config.VirtioFsNew("/fs1", "") }()
		fs2, _ := func() (config.VirtioDevice, error) { return config.VirtioFsNew("/fs2", "foo") }()
		net, _ := func() (config.VirtioDevice, error) { return config.VirtioNetNew("4a:f7:a9:77:38:d8") }()
		net2 := &config.VirtioNet{
			UnixSocketPath: "/tmp/unix.sock",
		}
		f, _ := os.CreateTemp(os.TempDir(), "tmp")
		net3 := &config.VirtioNet{
			Socket:     f,
			MacAddress: []byte{0x00, 0x11, 0x22, 0x33, 0x44, 0x55},
		}
		rng, _ := func() (config.VirtioDevice, error) { return config.VirtioRngNew() }()
		serial, _ := func() (config.VirtioDevice, error) { return config.VirtioSerialNew("/serial") }()
		vsock, _ := func() (config.VirtioDevice, error) { return config.VirtioVsockNew(1234, "/vsock", false) }()
		vsock2, _ := func() (config.VirtioDevice, error) { return config.VirtioVsockNew(1235, "/vsock2", true) }()
		input, _ := func() (config.VirtioDevice, error) { return config.VirtioInputNew(config.VirtioInputPointingDevice) }()
		input2, _ := func() (config.VirtioDevice, error) { return config.VirtioInputNew(config.VirtioInputKeyboardDevice) }()
		gpu, _ := func() (config.VirtioDevice, error) { return config.VirtioGPUNew() }()
		gpu2 := &config.VirtioGPU{
			UsesGUI: true,
			VirtioGPUResolution: config.VirtioGPUResolution{
				Width:  100,
				Height: 200,
			},
		}

		devs = append(devs,
			usb,
			usb2,
			blk,
			blk2,
			rosetta,
			nvme,
			nvme2,
			fs,
			fs2,
			net,
			net2,
			net3,
			rng,
			serial,
			vsock,
			vsock2,
			input,
			input2,
			gpu,
			gpu2,
		)

		x := Inspect(devs)
		c.JSON(200, x)
	})

	eng.Run(":7777")
}
```

**Result:**

> `curl http://127.0.0.1:7777/inspect`

```json
{
  "cpus": 0,
  "memory": 0,
  "devices": {
    "input": [
      {
        "kind": "virtioinput",
        "inputType": "pointing"
      },
      {
        "kind": "virtioinput",
        "inputType": "keyboard"
      }
    ],
    "gpu": [
      {
        "kind": "virtiogpu",
        "usesGUI": false,
        "width": 800,
        "height": 600
      },
      {
        "kind": "virtiogpu",
        "usesGUI": true,
        "width": 100,
        "height": 200
      }
    ],
    "vsock": [
      {
        "kind": "virtiosock",
        "port": 1234,
        "socketURL": "/vsock",
        "listen": false
      },
      {
        "kind": "virtiosock",
        "port": 1235,
        "socketURL": "/vsock2",
        "listen": true
      }
    ],
    "blk": [
      {
        "kind": "virtioblk",
        "devName": "virtio-blk",
        "imagePath": "/a",
        "readOnly": false,
        "deviceIdentifier": ""
      },
      {
        "kind": "virtioblk",
        "devName": "virtio-blk",
        "imagePath": "/b",
        "readOnly": false,
        "deviceIdentifier": "/vdd"
      }
    ],
    "fs": [
      {
        "kind": "virtiofs",
        "mountTag": "",
        "sharedDir": "/fs1"
      },
      {
        "kind": "virtiofs",
        "mountTag": "foo",
        "sharedDir": "/fs2"
      }
    ],
    "rosetta": {
      "mountTag": "/rosetta",
      "installRosetta": false
    },
    "nvme": [
      {
        "kind": "nvme",
        "devName": "nvme",
        "imagePath": "/nvme1",
        "readOnly": false
      },
      {
        "kind": "nvme",
        "devName": "nvme",
        "imagePath": "/nvme2",
        "readOnly": false
      }
    ],
    "net": [
      {
        "nat": true,
        "macAddress": "4a:f7:a9:77:38:d8",
        "unixSocketPath": "",
        "fd": 0
      },
      {
        "nat": false,
        "macAddress": "",
        "unixSocketPath": "/tmp/unix.sock",
        "fd": 0
      },
      {
        "nat": false,
        "macAddress": "00:11:22:33:44:55",
        "unixSocketPath": "",
        "fd": 9
      }
    ],
    "rng": true,
    "serial": {
      "logFile": "/serial",
      "usesStdio": false
    },
    "usbMassStorage": [
      {
        "kind": "usbmassstorage",
        "devName": "usb-mass-storage",
        "imagePath": "/usb1",
        "readOnly": false
      },
      {
        "kind": "usbmassstorage",
        "devName": "usb-mass-storage",
        "imagePath": "/usb2",
        "readOnly": false
      }
    ]
  }
}
```

### CODE REVIEW

File: pkg/rest/vf/vm_config.go [hunk 1]
```diff
@@ -4,28 +4,30 @@ import (
...
 	"github.com/gin-gonic/gin"
 	"github.com/sirupsen/logrus"
 )
 
 type VzVirtualMachine struct {
-	VzVM   *vz.VirtualMachine
-	config *vz.VirtualMachineConfiguration
+	VzVM     *vz.VirtualMachine
+	config   *vz.VirtualMachineConfiguration
+	vmConfig *config.VirtualMachine
```
Comment (cfergeau) [inline pkg/rest/vf/vm_config.go:16]: That's getting confusing, too many configs/vms/... data types ^^
Not a problem with this PR, just thinking out loud. Maybe will be reworked another day in a refactor

File: pkg/rest/vf/vm_config.go [hunk 2]
```diff
@@ -4,28 +4,30 @@ import (
...
 // and devices
 func (vm *VzVirtualMachine) Inspect(c *gin.Context) {
 	ii := define.InspectResponse{
-		// TODO complete me
-		CPUs:   1,
-		Memory: 2048,
-		//Devices: vm.Devices,
+		CPUs:   vm.vmConfig.Vcpus,
+		Memory: vm.vmConfig.MemoryBytes,
+		// TODO: add devices (VirtioDevice type needs to be improved) @Black-Hole1
```
Comment (cfergeau) [inline pkg/rest/vf/vm_config.go:29]: PR looks good to me without this added `TODO`

**Review by cfergeau — COMMENTED** [review]
Looks fine if you remove this `@Black-Hole1` comment.

File: pkg/rest/vf/vm_config.go (hunk 2 above)
Comment (cfergeau) [reply pkg/rest/vf/vm_config.go:29]: Great :) was not sure if this was now waiting for feedback from me, or if it was still work in progress. Great to know you are going to improve it some more!

File: pkg/rest/vf/vm_config.go (hunk 1 above)
Comment (BlackHole1) [reply pkg/rest/vf/vm_config.go:17]: Perhaps we can use the built-in http package in golang to handle this🤔. We can create an issue for this (and maybe add the `good first issue` label?).

**Comment by cfergeau** [comment]
> **Test Code:**

For what it's worth, if you want to add unit tests for this new code, it's now easier using https://github.com/crc-org/vfkit/blob/main/test/vm_test.go ,this makes it easy to create a VM with a set of devices, start vfkit, run some tests (eg make http request to the REST API), and check that the result is the expected one.

File: pkg/config/json_test.go [hunk 3]
```diff
@@ -56,7 +56,7 @@ var jsonTests = map[string]jsonTest{
 			require.NoError(t, err)
 			return vm
 		},
-		expectedJSON: `{"vcpus":3,"memoryBytes":4000000000,"bootloader":{"kind":"linuxBootloader","VmlinuzPath":"/vmlinuz","KernelCmdLine":"/initrd","InitrdPath":"console=hvc0"},"devices":[{"kind":"virtioblk","DevName":"virtio-blk","ImagePath":"/virtioblk1","ReadOnly":false,"DeviceIdentifier":""},{"kind":"virtioblk","DevName":"virtio-blk","ImagePath":"/virtioblk2","ReadOnly":false,"DeviceIdentifier":"virtio-blk2"}]}`,
+		expectedJSON: `{"vcpus":3,"memoryBytes":4000000000,"bootloader":{"kind":"linuxBootloader","VmlinuzPath":"/vmlinuz","KernelCmdLine":"/initrd","InitrdPath":"console=hvc0"},"devices":[{"kind":"virtioblk","devName":"virtio-blk","imagePath":"/virtioblk1","readOnly":false,"deviceIdentifier":""},{"kind":"virtioblk","devName":"virtio-blk","imagePath":"/virtioblk2","readOnly":false,"deviceIdentifier":"virtio-blk2"}]}`,
 	},
```
Comment (cfergeau) [inline pkg/config/json_test.go:60]: I need to check with podman devs if changing the json format will cause breakage for them. If it does, maybe we can add some compat code to `unmarshalDevice` to deal with the 2 formats, as I agree it looks nicer/more consistent/... to have lower case first letter.

File: pkg/rest/define/config.go [hunk 4]
```diff
@@ -1,11 +1,36 @@
...
 
+import (
+	"github.com/crc-org/vfkit/pkg/config"
+)
+
+type VirtioNetResponse struct {
+	Nat            bool   `json:"nat"`
+	MacAddress     string `json:"macAddress"`
+	UnixSocketPath string `json:"unixSocketPath"`
+	Fd             int    `json:"fd"`
```
Comment (cfergeau) [inline pkg/rest/define/config.go:11]: Fwiw, I'm not sure `fd` is going to be useful to expose to an external process?

File: pkg/rest/define/config.go [hunk 5]
```diff
@@ -1,11 +1,36 @@
...
+	GPU            []config.VirtioGPU            `json:"gpu"`
+	Vsock          []config.VirtioVsock          `json:"vsock"`
+	Blk            []config.VirtioBlk            `json:"blk"`
+	FS             []config.VirtioFs             `json:"fs"`
+	Rosetta        config.RosettaShare           `json:"rosetta"`
+	NVMe           []config.NVMExpressController `json:"nvme"`
+	Net            []VirtioNetResponse           `json:"net"`
+	Rng            bool                          `json:"rng"`
+	Serial         config.VirtioSerial           `json:"serial"`
+	USBMassStorage []config.USBMassStorage       `json:"usbMassStorage"`
```
Comment (cfergeau) [inline pkg/rest/define/config.go:25]: Can we reuse the same format as the preexisting json serialization code?
```
"devices":[{"devName":"virtio-blk","imagePath":"/virtioblk1","readOnly":false,"deviceIdentifier":""}]}
```
serialization/deserialization to this format should already be working, a new type should not be needed.
Or do you have strong reasons for having such a struct?

**Review by cfergeau — COMMENTED** [review]
This PR is going to be helpful to https://github.com/crc-org/vfkit/issues/48 to provide a way to let users of vfkit know which PTY was allocated :)

File: pkg/config/json_test.go (hunk 3 above)
Comment (cfergeau) [reply pkg/config/json_test.go:60]: podman devs are currently working towards a 5.0 release where breakage is acceptable, so this is a good opportunity to change this :)

**Comment by cfergeau** [comment]
> > **Test Code:**
> 
> For what it's worth, if you want to add unit tests for this new code, it's now easier using https://github.com/crc-org/vfkit/blob/main/test/vm_test.go ,this makes it easy to create a VM with a set of devices, start vfkit, run some tests (eg make http request to the REST API), and check that the result is the expected one.

I've started doing this in https://github.com/cfergeau/vfkit/commit/4615740c4b96d99b7491dee8481246272c01108c but not finished yet :)

File: pkg/rest/define/config.go (hunk 5 above)
Comment (cfergeau) [reply pkg/rest/define/config.go:25]: https://github.com/crc-org/vfkit/pull/99 reuses the existing json format

**Comment by cfergeau** [comment]
> I've started doing this in [cfergeau@4615740](https://github.com/cfergeau/vfkit/commit/4615740c4b96d99b7491dee8481246272c01108c) but not finished yet :)

I'm now done with the testing code/..., I've filed https://github.com/crc-org/vfkit/pull/99 , in my opinion we can move the rest of the discussions in this new PR.

**Comment by BlackHole1** [comment]
Hey @cfergeau, because it's Chinese New Year now, I'm currently on vacation and will be back in about a week.

**Comment by cfergeau** [comment]
> Hey @cfergeau, because it's Chinese New Year now, I'm currently on vacation and will be back in about a week.

Thanks for the message, enjoy the holidays!

File: pkg/rest/vf/vm_config.go (hunk 1 above)
Comment (cfergeau) [reply pkg/rest/vf/vm_config.go:17]: I've started reworking this in this branch https://github.com/cfergeau/vfkit/commits/vzvm/ as it seemed to be a requirement for other changes I wanted to make, see the `rework types` commit.
The idea is to have a `VfVirtualMachineConfig` type which wraps `vz.VirtualMachineConfiguration` + `vfkit.config.VirtualMachine` and a `VfVirtualMachine` type which wraps `vz.VirtualMachine` and `VfVirtualMachineConfig`.

This way, with only a `VfVirtualMachine` instance, you have access to all the `vz` and `vf` related objects.

**Comment by BlackHole1** [comment]
I'm back :)

> I'm now done with the testing code/..., I've filed https://github.com/crc-org/vfkit/pull/99 , in my opinion we can move the rest of the discussions in this new PR.

Does it look like I can close this PR?

**Comment by cfergeau** [comment]
> I'm back :)
> 
> > I'm now done with the testing code/..., I've filed #99 , in my opinion we can move the rest of the discussions in this new PR.
> 
> Does it look like I can close this PR?

If the json format proposed in #99 for the `inspect` API works for you, then yes we can close this PR.

**Comment by BlackHole1** [comment]
> If the json format proposed in https://github.com/crc-org/vfkit/pull/99 for the inspect API works for you

As for me, the new JSON format will make Go developers comfortable, but I cannot be certain about developers in other languages because they may not be able to determine the result of the response by inspecting the source code. Perhaps we still need to improve the relevant documentation?
I am also concerned that the new format may introduce more subtle breaking changes. For example, if I modify or delete a key but am unaware that this key is reflected in the `inspect API`, it could lead to breaking changes inadvertently.

**Comment by cfergeau** [comment]
> Perhaps we still need to improve the relevant documentation?

Ah yes, for sure, I did not even think about documenting this yet :-/

> For example, if I modify or delete a key but am unaware that this key is reflected in the inspect API, it could lead to breaking changes inadvertently.

`pkg/config/json_test.go` should catch many of these breaking changes, but it's not exhaustive, we could extend it to make sure it covers all the fields that are serialized to json. This way json breaking changes will be noticeable.

**Comment by cfergeau** [comment]
> > Perhaps we still need to improve the relevant documentation?
> 
> Ah yes, for sure, I did not even think about documenting this yet :-/
> 
> > For example, if I modify or delete a key but am unaware that this key is reflected in the inspect API, it could lead to breaking changes inadvertently.
> 
> `pkg/config/json_test.go` should catch many of these breaking changes, but it's not exhaustive, we could extend it to make sure it covers all the fields that are serialized to json. This way json breaking changes will be noticeable.

I tried to do this in https://github.com/crc-org/vfkit/pull/103

**Pull request closed without merging**